### ISSUE TITLE
Different number of masked positions  of  samples in  the same batch  for RoBERTa

### ISSUE THREAD

**Issue by zhaoguangxiang** [pytorch/fairseq#1699]
## ❓ Questions and Help
In pretraining of Roberta, what is the effect of adding a random number to the number of masked positions( 0.15*seq_len +np.random.rand()) instead of fixed length for all samples in a batch?

e.g.
for a batch of 8 samples, the seq_len after padding is 339
masked_tokens.size:  torch.Size([8, 339])
masked_token_len tensor([50, 50, 51, 51, 51, 51, 51, 51], device='cuda:0')
but the number of masked tokens is not always 50 or 51
`
            num_mask = int(
                # add a random number for probabilistic rounding
                self.mask_prob * sz + np.random.rand()
            )
`

**Comment by lematt1991** [comment]
Please provide the full command you used to launch your training job.  What is your `mask_prob`?  Are you using `--mask-whole-words`?

**Comment by zhaoguangxiang** [comment]
mask_prob is 0.15,   do not use mask-whole-words. will mask-whole-words leads to the uneven number of mask_tokens even  without random number for probalilistic rounding

`
TOTAL_UPDATES=125000    # Total number of training steps
WARMUP_UPDATES=10000    # Warmup the learning rate over this many updates
PEAK_LR=0.0005          # Peak learning rate, adjust as needed
TOKENS_PER_SAMPLE=512   # Max sequence length
MAX_POSITIONS=512       # Num. positional embeddings (usually same as above)
MAX_SENTENCES=8        # Number of sequences per batch (batch size)
UPDATE_FREQ=32          # Increase the batch size 16x

DATA_DIR=data-bin/wikitext-103
mkdir checkpoint
cur_save=roberta_wiki103
python3 train.py --fp16 $DATA_DIR \
    --task masked_lm --criterion masked_lm \
    --arch roberta_base --sample-break-mode complete --tokens-per-sample $TOKENS_PER_SAMPLE \
    --optimizer adam --adam-betas '(0.9,0.98)' --adam-eps 1e-6 --clip-norm 0.0 \
    --lr-scheduler polynomial_decay --lr $PEAK_LR --warmup-updates $WARMUP_UPDATES --total-num-update $TOTAL_UPDATES \
    --dropout 0.1 --attention-dropout 0.1 --weight-decay 0.01 \
    --max-sentences $MAX_SENTENCES --update-freq $UPDATE_FREQ  --max-update $TOTAL_UPDATES --ddp-backend=no_c10d \
    --log-format json --tensorboard-logdir checkpoint/${cur_save}  --save-dir checkpoint/${cur_save} --save-interval 10  2>&1 | tee checkpoint/${cur_save}.txt
`

**Comment by myleott** [comment]
I doubt it makes a big difference. We're just trying to keep the percentage of masked tokens close to 0.15, but if you do `num_mask = int(self.mask_prob * sz)` then you'll always be rounding down. For example if `sz=50` then you'll always mask 7 tokens, which is 14% of tokens. By adding the random number we'll sometimes pick 7 and sometimes pick 8, which on average will yield 15% masks.

**Comment by zhaoguangxiang** [comment]
> I doubt it makes a big difference. We're just trying to keep the percentage of masked tokens close to 0.15, but if you do `num_mask = int(self.mask_prob * sz)` then you'll always be rounding down. For example if `sz=50` then you'll always mask 7 tokens, which is 14% of tokens. By adding the random number we'll sometimes pick 7 and sometimes pick 8, which on average will yield 15% masks.

so it is equal to  `num_mask = int(self.mask_prob*sz +0.5)`  ?

**Comment by myleott** [comment]
No, because then for `sz=50` it would always mask 8 tokens (=16%). Here are some details about stochastic rounding: https://en.wikipedia.org/wiki/Rounding#Stochastic_rounding
